File: tests/fixtures/check_then_act/ts/transaction_boundary_fixture.ts
// A check outside the unit of work guards a write inside it; the second method re-checks inside.
export class Builder {
  constructor(private readonly db: any, private readonly repo: any) {}

  build(key: string) {
    const found = this.repo.find(key);
    if (found) return found; // expect: check
    return this.db.unitOfWork(() => { // expect: gap transaction
      return this.repo.insert(key); // expect: write
    });
  }

  buildRechecked(key: string) {
    const found = this.repo.find(key);
    if (found) return found;
    return this.db.unitOfWork(() => {
      if (this.repo.find(key)) return undefined;
      return this.repo.insert(key);
    });
  }
}

File: tests/fixtures/check_then_act/py/transaction_boundary_fixture.py
# A check outside the transaction guards a write inside it; the second method re-checks inside.
from django.db import transaction


class Builder:
    def __init__(self, repo):
        self.repo = repo

    def build(self, key):
        found = self.repo.find(key)
        if found:  # expect: check
            return found
        with transaction.atomic():  # expect: gap transaction
            return self.repo.insert(key)  # expect: write

    def build_rechecked(self, key):
        found = self.repo.find(key)
        if found:
            return found
        with transaction.atomic():
            if self.repo.find(key):
                return None
            return self.repo.insert(key)
